### PR TITLE
Use shorter ID prefixes in configured revset

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by PhilipMetzger** [comment]
SGTM, besides the typo in the last sentence of the sixth commit message `I've update the templater to use it.` which should be `I've updated the templater to use it`. I'll defer to @yuja for all revset related changes though.

I do have a question related to the fourth commit as it will impact my PR(#1200). Is there a good pattern for the default `SymbolResolver` usage?

**Comment by martinvonz** [comment]
> I do have a question related to the fourth commit as it will impact my PR(#1200). Is there a good pattern for the default `SymbolResolver` usage?

I don't think it should impact your PR. Were you thinking that we'd implement `jj prev 5` by resolving "5" as a revset symbol pointing that many commits back? I think it's better to just parse it as an integer.

**Comment by PhilipMetzger** [comment]
No I didn't use the revset symbols, although it could've worked. 

But I restructured the PR to use `RevsetExpressions` recently, using this pattern: 
```rust
// Similar to cmd_next, on my machine 
let helper = ...;
let current_wc_id = helper.get_wc_id()?;
let parent = current_wc_id.parents().pop();
let all_ancestors = RevsetExpression::ancestors(&parent);
//...
```
which will need to use the Resolver.

**Comment by PhilipMetzger** [comment]
Thanks for clearing my concerns.

File: CHANGELOG.md
Comment (ilyagr) [inline CHANGELOG.md:87]: Re the commit message (for the last commit):

I might be missing something, since I'm not sure I fully understood the connection between the default value and unifying revsets and revset-aliases yet.

However, naively, the following seems to make more sense:

- Default `revsets.short-prefixes` to "" in rust code.
- In TOML, set its default value.
 
I would still prefer `(remote_branches(main)|remote_branches(master)|remote_branches(develop)|tags())..` as the default value.

For `revset.log`, I would actually make the default to be `all()` in Rust and something else in TOML.

As an aside, perhaps the noop value for `revsets.short-prefixes` should be `none()`?

Comment (ilyagr) [reply CHANGELOG.md:87]: Regarding the default for `revsets.short-prefixes`, one motivation is that it's pretty weird to see the revsets grow when you do `jj git push --all`.

My suggested default for short-prefixes might also be a decent default for log (though it should also include the parents of those commits).

I don't mean to block the PR on figuring out the defaults, though.

-----------------------------------

Also, here's a weird behavior I'm seeing with the default setting, right after `jj git push --all`. Why does `V` get a short id? (If it's not obvious to you, I'll try to get more detailed debug info).

image

Comment (martinvonz) [reply CHANGELOG.md:87]: > Why does V get a short id?

It looks like it's the "base" for your `X` commit. It should show up in plain `jj log` for the same reason.

Comment (martinvonz) [reply CHANGELOG.md:87]: > Re the commit message (for the last commit):
> 
> I might be missing something, since I'm not sure I fully understood the connection between the default value and unifying revsets and revset-aliases yet.

What I meant was that if `[revset-aliases]` gets merged into `[revsets]`, then the user can define `revsets.short-prefixes = "log"` and `revsets.log = "..."`, but if they're not merged, then the user cannot do `revsets.short-prefixes = "log"`.

> However, naively, the following seems to make more sense:
> 
> * Default `revsets.short-prefixes` to "" in rust code.
> * In TOML, set its default value.

Yes, it might be a little cleaner to put the defaults in TOML so most defaults are collected in one place, and so they show up in `jj config list --include-defaults`.

> I would still prefer `(remote_branches(main)|remote_branches(master)|remote_branches(develop)|tags())..` as the default value.

Do you mean without the "base" commits or did you mean to include them? I may I misunderstood you on Discord because I thought you said you want short hashes for the base commits too.

> For `revset.log`, I would actually make the default to be `all()` in Rust and something else in the default TOML config.
> 
> As an aside, perhaps the noop value for `revsets.short-prefixes` should be `none()`?

I agree that we should move configs to TOML where possible.

File: lib/src/revset.rs
Comment (yuja) [inline lib/src/revset.rs:1646]: nit: `{symbol:?}`

File: src/cli_util.rs
Comment (yuja) [inline src/cli_util.rs:957]: Boxing isn't needed here.

File: CHANGELOG.md
Comment (ilyagr) [reply CHANGELOG.md:87]: > Do you mean without the "base" commits or did you mean to include them? 

I don't feel very strongly about it, but I would prefer to include the "base" commits in the `short-prefixes` default. I just forgot to mention them.

Including "base" commits seems more useful in the `log` default. But then, there are more changes I'd like to make to the `log` default. For example, I think it'd also be nice for that to include the tips of local branches.

File: lib/src/revset.rs
Comment (yuja) [inline lib/src/revset.rs:1651]: This can be of `Fn(&dyn Repo, &HexPrefix) -> ...` type. Then `&dyn Repo` reference can be removed from `IdPrefixContext` and `unsafe {}` cast in CLI will go away.

A downside is that the caller needs to guarantee the `repo` reference passed to `IdPrefixContext` is unchanged, but I don't think that's worse than the current design. We'll anyway need to ensure that `IdPrefixContext` is initialized with the same `repo` as the one `DefaultSymbolResolver` holds.

File: lib/src/id_prefix.rs
Comment (yuja) [inline lib/src/id_prefix.rs:152]: nit: typo in commit message: s/IdPrefix/IdIndex/
> prefixes: move `IdPrefix` to `id_prefix` module
> id_prefix: add `IdPrefix::resolve_prefix()`

File: lib/src/id_prefix.rs
Comment (yuja) [inline lib/src/id_prefix.rs:242]: `.unwrap_or(1)` ?

If the index only contained the `key`, both `left` and `right` would be `None`.

File: src/cli_util.rs
Comment (yuja) [inline src/cli_util.rs:953]: Borrowing `workspace_id` here makes the unsafe cast scary because it's no longer a self-referential struct.

Perhaps, it's better to make `DisambiguationData` own `Option<WorkspaceId>`.

File: lib/src/id_prefix.rs
Comment (yuja) [inline lib/src/id_prefix.rs:40]: nit: `repo: &dyn Repo` (the lifetime `'a` isn't required)

**Review by yuja — APPROVED** [review]
Nice

File: CHANGELOG.md
Comment (martinvonz) [reply CHANGELOG.md:87]: I'll leave this for later to avoid blocking this PR on what seems like a larger discussion. I can't even move the default value to TOML in this PR since the default is "same as revsets.log", which can't be expressed in TOML unless we merge `[revsets]` and `[revset-aliases]`.

File: src/cli_util.rs
Comment (martinvonz) [reply src/cli_util.rs:949]: Yeah, that's much better. Thanks.

File: lib/src/revset.rs
Comment (martinvonz) [reply lib/src/revset.rs:1650]: FYI, I'm looking into this now.

Comment (martinvonz) [reply lib/src/revset.rs]: I made that change in a commit on top. Let me know if you prefer that I absorb it into the other commits.

File: src/cli_util.rs
Comment (yuja) [reply src/cli_util.rs:952]: I meant the concrete type `-> DefaultSymbolResolver<'_>`, but `impl` should also be fine.

File: lib/src/revset.rs
Comment (yuja) [reply lib/src/revset.rs]: LGTM, thanks!

File: src/cli_util.rs
Comment (martinvonz) [reply src/cli_util.rs:952]: I switched it to `DefaultSymbolResolver<'_>`

**Pull request merged**
